text/0481-component-templates-co-location.md

```diff
@@ -0,0 +1,647 @@
+
+```js
+// addon/components/foo-bar.js
+
+import Component from '@ember/component';
+import layout from '../templates/components/foo-bar';
+
+export default Component.extend({
+  layout
+});
+```
+
+```hbs
+{{!-- addon/templates/components/foo-bar.hbs --}}
+
+{{yield}}
+```
+
+```js
+// app/components/foo-bar.js
+
+export { default } from 'my-addon/components/foo-bar';
+```
+
+First, within the `addon` folder, there is the component class and template.
+Having these files in the `addon` folder allow them to be imported from
+JavaScript (e.g. `import FooBar from 'my-addon/components/foo-bar';`) for
+subclassing. However, as mentioned above, this does not also bring along the
+component's template.
+
+To avoid developers having to also import the template, the addon component
+blueprint sets the `layout` property on the component's class. At runtime,
+when Ember fail to find a template associated with the component, it will
+then fall back to this property.
+
+Finally, it re-exports the component in the `app` folder, which is merged with
+the app's `app` folder, allowing the component to be resolved by the resolver
+and invoked from Handlebars.
+
+This workaround gets the job done, but has several major drawbacks.
+
+First, it adds a significant amount of boilerplate to addon components, which
+could be puzzling even to seasoned addon developers. In fact, the problem it
+solves is unique to addons – there is nothing wrong with using inheritance in
+apps – we just didn't think we could justify adding this amount of boilerplate
+to every app and putting them in Ember's "happy path".
+
+Second, the `layout` property in classic component uses a feature in Glimmer VM
+called "late-bound layout", in which we must assume a different template for
+each _instance_ of the component (the `layout` property can be set in `init`).
+This is fundamentally more difficult to optimize and excludes the component
+from key optimizations (and opportunities for future optimizations).
+
+Most importantly, Glimmer components do not have the equivalent of the `layout`
+property, meaning that addons (or components designed for inheritance in mind)
+cannot be written using Glimmer components today.
+
+To solve these problems once and for all, we need to change things such that
+component templates are automatically associated with their JavaScript class at
+build time, and importing the JavaScript class should result in a value that
+has this metadata associated with it, which needs to survive subclassing as
+well. Solving this problem is also essential for unblocking template imports.
+
+We believe the technical design for allowing co-location will solve this
+problem nicely.
+
+## Detailed design
+
+### High-level design
+
+We propose to allow placing a component's template adjacent to its JavaScript
+file in `app/components`. For example, for a component named `foo-bar`, it will
+be `app/components/foo-bar.js` and `app/components/foo-bar.hbs`.
```

Comment:
I have some concerns about supporting components in a folder (`app/components/foo-bar/index.js`) and also components at the top level of the components directory (`app/components/foo-bar.js`).

Which of these structures will the CLI default to? Will there be a config setting to enforce usage of one form over the other on a project-wide basis?

For my own projects, I would personally prefer the folder based approach, so I do appreciate the flexibility but it seems like having two file structure options reduces the consistency and could lead to developer confusion.

In particular, the non-folder based structure seems like it could quickly create a mess of the components directory, especially if other file types are introduced, via further folder structure refinements or addons. Although the following component files are located together, it's not particularly easy to parse what belongs to what.

```
- components
  - first-component.js
  - first-component.hbs
  - first-component.css
  - first-component-tests.js
  - first-component-other.js
  - second-component.js
  - second-component.hbs
  - second-component.css
  - second-component-tests.js
  - second-component-other.js
```

Reply:
I personally also much prefer the "pods" approach, exactly for the reason that it is easier to extend, e.g. `styles.css` files or even further "private" files that are imported by the `component.js`.